migrate to eslint & update dependencies

boranx opened the pull request:
Fix: https://github.com/kubernetes/test-infra/issues/25896

execution of the validations:
```shell
$ make verify-eslint
hack/make-rules/verify/eslint.sh
yarn install v1.22.19
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.22s.

.
.
✖ 3168 problems (3024 errors, 144 warnings)
  2690 errors and 0 warnings potentially fixable with the `--fix` option.
```

chaodaiG commented:
/ok-to-test

chaodaiG commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-verify-lint/1552297835116892160, lot of the errors are indentation error, searched "indentation" in the log and found 2568 occurrence, is there a tool to auto correct them?

boranx commented on prow/cmd/webhook-server/main.go:
```diff
@@ -35,10 +35,10 @@ import (
-	caCert = "ca-cert"
-	caPrivKey = "ca-priv-key"
-	caBundle = "ca-bundle"
+	secretID       = "prowjob-webhook-ca-cert"
```
those are probably leftovers from go-lint after I executed `make`, removed :+1:

boranx commented:
@chaodaiG passing `--fix` flag to `eslint` makes them resolved[1] but it creates diff as they are not committed
I think that should be done local and then needs to be pushed

1- e.g: `node_modules/eslint/bin/eslint -c eslintrc.js --fix --ext .ts .`

chaodaiG commented:
sounds good, could you push the diff in a new commit so that it's easier to review?

boranx commented:
ack, the option `--fix` has been included and the diff added.

boranx commented:
/retest

chaodaiG commented on hack/run-in-node-container.sh:
```diff
@@ -22,8 +22,8 @@ set -o pipefail
 
-# NOTE: bazel is currently on node 12.x, but we should probably upgrade at some point
-NODE_IMAGE='node:12-bullseye-slim'
+# NOTE: bazel is currently on node 14.x, but we should probably upgrade at some point
```
this comment is irrelevant any more

boranx replied:
ack, removed

chaodaiG commented on package.json:
```diff
@@ -18,17 +18,21 @@
     "rollup": "2.3.4",
     "rollup-plugin-amd": "4.0.0",
+    "rollup-plugin-terser": "5.1.3",
     "terser": "4.4.0",
```
I might be wrong but iirc this is what we use, should we also bump it?

boranx replied:
I'm not sure, I've updated this to the latest anyway

chaodaiG commented on package.json:
```diff
@@ -18,17 +18,21 @@
+    "eslint-plugin-jsdoc": "^39.3.3",
+    "eslint-plugin-prefer-arrow": "^1.2.3",
     "jasmine": "~3.4.0",
     "rollup": "2.3.4",
```
do we also want to bump this version?

boranx replied:
again for this one, I've updated to the latest

boranx commented:
@chaodaiG if I add the diff, the CI fails: 
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-integration/1552343733867909120
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-prow-image-build-test/1552343733909852160
any idea why that might be happening?

btw I've reverted the changes back due to CI issue, wfm on local for `make verify-eslint`

chaodaiG commented:
> @chaodaiG if I add the diff, the CI fails: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-integration/1552343733867909120 https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-prow-image-build-test/1552343733909852160 any idea why that might be happening?
> 
> btw I've reverted the changes back due to CI issue, wfm on local for `make verify-eslint`

these tests both compile typescripts as part of the prow images building, the typescript change you made probably didn't work with the typescript toolchain versions that are used in this repo. You'll need to update them according to the comments that I left on your PR. These tools https://github.com/kubernetes/test-infra/blob/922f2b0d911f48fc4c38ce86f4b4aa3d5f6aac83/hack/ts-rollup/main.go#L40-L43 in particular are used in typescript compilation, and their versions are controlled by package.json

alvaroaleman commented:
/label tide/merge-method-squash

chaodaiG commented:
The typescript compilation in prow integration tests and image build tests are triggered via https://github.com/kubernetes/test-infra/blob/master/prow/cmd/deck/gather-static.sh, as it's invoked from Go program, it's a bit hard to read its output. I think it will be easier to debug the type script compilation failure by directly invoke this script from your terminal:
```
$ ./prow/cmd/deck/gather-static.sh
```

boranx commented:
hey @chaodaiG 

I think I found the issue, applying the changes from `--fix` causing the integration tests to be failing for the following files:

```git
        modified:   prow/cmd/deck/static/command-help/command-help.ts
        modified:   prow/cmd/deck/static/pr/pr.ts
```
so I've reverted changes on these two files back in the hope to see the pipelines pass

boranx commented:
yeah so there's only linting job left as failing, that's because there are issues reported by linter:

```shell
✖ 1667 problems (1544 errors, 123 warnings)
  1200 errors and 0 warnings potentially fixable with the `--fix` option.

make: *** [Makefile:68: verify-eslint] Error 1
```

chaodaiG commented:
> yeah so there's only linting job left as failing, that's because there are issues reported by linter:
> 
> ```shell
> ✖ 1667 problems (1544 errors, 123 warnings)
>   1200 errors and 0 warnings potentially fixable with the `--fix` option.
> 
> make: *** [Makefile:68: verify-eslint] Error 1
> ```

This seems expected to me. Our typescript toolchains versions, as you probably have seen, are pretty ancient, and the goal here is to bring them up-to-date. Even though the issue says update the linter for typescript, I can totally imagine that updated linter would conflict with the rest of the typescript compilation tools because they are too old. And just bump the version of the typescript compilation tools might not be sufficient to fix all of the compilation errors, as there might be some usages/patterns not caught by linters, and unfortunately they might need to be fixed manually, which can be started by running the `gather-statics.sh` script as I mentioned in the previous comment.

boranx commented:
indeed, but that sounds more like refactoring and needs to be worked on more so it might worth to be issuing this as a new item to work on suggestions on linter/refactoring...

I'm not sure I can cover all reported issues in this PR in a short-term time period as there are 461 problems (343 errors, 118 warnings) to be fixed

chaodaiG commented:
> indeed, but that sounds more like refactoring and needs to be worked on more so it might worth to be issuing this as a new item to work on suggestions on linter/refactoring...
> 
> I'm not sure I can cover all reported issues in this PR in a short-term time period as there are 461 problems (343 errors, 118 warnings) to be fixed

fair enough! Would you please create a separate issue as a feature request of refactoring typescripts and log all your findings there?

boranx commented:
sure! https://github.com/kubernetes/test-infra/issues/26959 has been created to track the issues reported by the linter

boranx commented:
/test pull-test-infra-integration

boranx commented:
hey @chaodaiG,

I found some time today to look at this PR and fix the issues, I've applied some of the suggestions, and some of them (the most reoccurring ones) transformed to be printed as warn, see the `eslintrc.js`

```
        "@typescript-eslint/require-await": "warn",
        "@typescript-eslint/no-floating-promises": "warn",
        "@typescript-eslint/no-unsafe-call": "warn",
        "@typescript-eslint/no-unsafe-return": "warn",
        "@typescript-eslint/no-unsafe-argument": "warn",
        "@typescript-eslint/no-unsafe-assignment": "warn",
        "@typescript-eslint/no-misused-promises": "warn",
``` 

Eventually, unblocking rules one by one and fixing them might be a way of gracefully making changes instead of a bulk change. Currently, all are being reported as warnings and doesn't break anything on CI as far as I can see:

```shell
✖ 386 problems (0 errors, 386 warnings)
```

chaodaiG commented:
Understand that it's a big ask to do two things in one PR: update lint + update typescript compilation tools. This PR at least updated to eslint and the node version, so this LGTM. However, I would prefer to separate the generated `eslintrc.js` with your update into two separate PRs to makes it more clear to future readers, namely:

```
        "@typescript-eslint/require-await": "warn",
        "@typescript-eslint/no-floating-promises": "warn",
        "@typescript-eslint/no-unsafe-call": "warn",
        "@typescript-eslint/no-unsafe-return": "warn",
        "@typescript-eslint/no-unsafe-argument": "warn",
        "@typescript-eslint/no-unsafe-assignment": "warn",
        "@typescript-eslint/no-misused-promises": "warn",
```

should be in a separate commit. Could you squash this PR to include two commits? the first one is almost everything except for any manual updates in `eslintrc.js` file

boranx commented:
ack, understood. I've gathered them in 2 commits, the second one just reflects the `eslint` tweak required for ci

chaodaiG commented:
/retest

chaodaiG commented:
awesome! thank you! One more thing, would you be able to run https://github.com/kubernetes/test-infra/tree/master/prow/cmd/deck#running-deck-locally locally to make sure that the UI won't crash?

boranx commented:
works well on my local: http://localhost:8080/

chaodaiG commented:
great!
/label tide/merge-method-merge

chaodaiG commented:
/lgtm
/approve
/hold
let's merge it after the automated bump later today(~4:30PM PST)

chaodaiG reviewed (commented):
/unhold

chaodaiG commented:
/lgtm
/approve

chaodaiG commented:
/remove-label tide/merge-method-squash
this PR was squashed by the author already
